test/integration/tcp_proxy_integration_test.cc

```diff
@@ -0,0 +1,125 @@
+#include "test/integration/tcp_proxy_integration_test.h"
+
+#include "common/network/utility.h"
+#include "common/ssl/context_manager_impl.h"
+
+#include "test/integration/ssl_utility.h"
+#include "test/integration/utility.h"
+#include "test/mocks/runtime/mocks.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+INSTANTIATE_TEST_CASE_P(IpVersions, TcpProxyIntegrationTest,
+                        testing::ValuesIn(TestEnvironment::getIpVersionsForTest()));
+
+TEST_P(TcpProxyIntegrationTest, TcpProxyUpstreamDisconnect) {
+  IntegrationTcpClientPtr tcp_client;
+  FakeRawConnectionPtr fake_upstream_connection;
+  FakeRawConnectionPtr fake_rest_connection;
+  executeActions({
+      [&]() -> void { tcp_client = makeTcpConnection(lookupPort("tcp_proxy")); },
+      [&]() -> void { tcp_client->write("hello"); },
+      [&]() -> void { fake_upstream_connection = fake_upstreams_[0]->waitForRawConnection(); },
+      [&]() -> void { fake_upstream_connection->waitForData(5); },
+      [&]() -> void { fake_upstream_connection->write("world"); },
+      [&]() -> void { fake_upstream_connection->close(); },
+      [&]() -> void { fake_upstream_connection->waitForDisconnect(); },
+      [&]() -> void { tcp_client->waitForDisconnect(); },
+
+      // Clean up unused client_ssl_auth
+      [&]() -> void { fake_rest_connection = fake_upstreams_[1]->waitForRawConnection(); },
+      [&]() -> void { fake_rest_connection->close(); },
+      [&]() -> void { fake_rest_connection->waitForDisconnect(); },
+  });
+
+  EXPECT_EQ("world", tcp_client->data());
+}
+
+TEST_P(TcpProxyIntegrationTest, TcpProxyDownstreamDisconnect) {
```

Comment:
Can you add one liner comments above each `TEST_P` case to explain what's being checked?